### PR TITLE
fields can be used directly in inner builder

### LINKED ISSUE #11: Requires constructor already present
A fairly minor niggle in an otherwise awesome project:

It seems as though the builder requires that a constructor already be present.

When trying to use the builder with a class without a constructor, the "wizard" doesn't list any fields to use.

This makes the workflow:
 * Create a class
 * Create a constructor with a set of fields
 * Create a builder.

It'd be great not have to perform step 2.

### PR DESCRIPTION
**Pull request by blindpirate**
I just made some tiny improvements that user can use fields which donot appear in constructor/setter in case of inner builder. In that case, builder will assign value to field directly.

For example,

```
public class Foo {
    int a;

    public int getA() {
        return a;
    }
}
```

will be

```
public class Foo {
    int a;

    public int getA() {
        return a;
    }


    public static final class FooBuilder {
        int a;

        private FooBuilder() {
        }

        public static FooBuilder aFoo() {
            return new FooBuilder();
        }

        public FooBuilder withA(int a) {
            this.a = a;
            return this;
        }

        public Foo build() {
            Foo foo = new Foo();
            foo.a = this.a;
            return foo;
        }
    }
}
```
That helps me a lot because I want inner builder to be more convenient.
And this resolve #11 
Thank you very much.

### CODE REVIEW

**Comment by mjedynak** [comment]
When I applied your changes, the generated Foo class looks as follows:
```
public class Foo {
    int a;

    public int getA() {
        return a;
    }

    public static final class FooBuilder {
        private FooBuilder() {
        }

        public static FooBuilder aFoo() {
            return new FooBuilder();
        }

        public FooBuilder withA(int a) {
            this.a = a;
            return this;
        }

        public Foo build() {
            Foo foo = new Foo();
            foo.a = this.a;
            return foo;
        }
    }
}
```

Missing `int a` in builder causes compilation issues. 
Generally I'm happy with the enhancement and I'm willing to merge it once you resolve this issue.

File: .gitignore [hunk 1]
```diff
@@ -2,4 +2,7 @@
 *.ipr
 *.iws
 .idea
-.idea/*
\ No newline at end of file
+.idea/*
+build/*
+.gradle/*
+out/*
```
Comment (mjedynak) [inline .gitignore]: Idea ignores it by default, I don't think we need it.

File: README [hunk 2]
```diff
@@ -5,4 +5,5 @@ Requires gradle and IntelliJ IDEA for building.
 Build instructions:
 - edit build.gradle and change values of 'ideaInstallationPath' and 'ideaJdk' to be in sync with your environment
 - invoke 'gradle build' from command line to build
-- invoke 'gradle idea' from command line to generate IDEA specific files - provides possibility to be opened as IDEA project.
\ No newline at end of file
+- invoke 'gradle idea' from command line to generate IDEA specific files - provides possibility to be opened as IDEA project.
+- in IDEA, select 'File' - 'Open' - ${YOUR_DIR}/Builder-Generator.ipr, donot use 'File' - 'Import' !!!!
```
Comment (mjedynak) [inline README]: I've recently added a tip about the ipr file.

File: build.gradle [hunk 3]
```diff
@@ -6,6 +6,8 @@ apply plugin: 'checkstyle'
 // --- properties ---
 ext.ideaInstallationPath = 'c:/Program Files (x86)/JetBrains/IntelliJ IDEA Community Edition 2016.1.1'
 ext.ideaJdk = 'IntelliJ IDEA Community Edition IC-145.597.3'
+//ext.ideaInstallationPath = '/Applications/IntelliJ IDEA.app/Contents'
+//ext.ideaJdk = 'IntelliJ IDEA IU-162.1812.17'
 sourceCompatibility = 1.6
```
Comment (mjedynak) [inline build.gradle]: please do not commented stuff

File: src/test/java/pl/mjedynak/idea/plugins/builder/action/handler/DisplayChoosersRunnableTest.java [hunk 4]
```diff
@@ -40,26 +40,46 @@
...
-    @Mock private PsiFile psiFile;
-    @Mock private PsiDirectory psiDirectory;
-    @Mock private PsiPackage psiPackage;
-    @Mock private PsiManager psiManager;
-    @Mock private CreateBuilderDialog createBuilderDialog;
-    @Mock private MemberChooser memberChooserDialog;
-    @Mock private PsiFieldsForBuilderFactory psiFieldsForBuilderFactory;
-    @Mock private PsiFieldsForBuilder psiFieldsForBuilder;
+    @InjectMocks
+    private DisplayChoosersRunnable displayChoosersRunnable;
```
Comment (mjedynak) [inline src/test/java/pl/mjedynak/idea/plugins/builder/action/handler/DisplayChoosersRunnableTest.java]: please do not change the formatting

File: .gitignore (hunk 1 above)
Comment (blindpirate) [inline .gitignore]: That's werid. I am using Idea Ultimate on Mac OS 10.11, if I do not add those lines to my `.gitignore`, contents in those directory will not be ignored automatically. Am I wrong at somewhere? I'm not an experienced use of Idea.
image

**Comment by blindpirate** [comment]
@mjedynak The issue you mentioned was fixed in my last commit [2c94cd5](https://github.com/mjedynak/builder-generator-idea-plugin/pull/30/commits/2c94cd5c8e729febfbd59ce20ed37af0fc2b68b5), I'm afraid you didn't apply that change. Do you mind apply the commit and try again? 

And I will change the format back.

Thank you very much

File: README (hunk 2 above)
Comment (blindpirate) [inline README]: Yes, that's really necessary for people who are not familiar with plugin development, it takes me several hours to find the right way. Please ignore my modification.

File: .gitignore (hunk 1 above)
Comment (mjedynak) [inline .gitignore]: Not sure what's going on, let's add this to `.gitignore`

**Comment by mjedynak** [comment]
For such class:
```
public class Person {                    
                                         
    private final String name;           
    private final int age;               
                                         
    public Person(String name, int age) {
        this.name = name;                
        this.age = age;                  
    }
```          
it produces compilation error, because of final fields.
In such case it should create inner builder 'the old way' by using constructor.

Please resolve also the conflicts, so can merge it automatically.

**Comment by blindpirate** [comment]
I fixed the issue you metioned and resolved the conflict. It's so brilliant of you to find that edge case.

Please check if there are any other questions, I'm happy to be a contributor.

File: src/test/java/pl/mjedynak/idea/plugins/builder/psi/BuilderPsiClassBuilderTest.java [hunk 5]
```diff
@@ -223,6 +248,70 @@ public void shouldReturnBuilderObjectWithBuildMethodUsingSetterAndConstructor()
...
+    public void constructorShouldHavePriorityOverSetter() {
+        // given
+        PsiField nameField = mock(PsiField.class);
+        PsiField ageField = mock(PsiField.class);
+        given(nameField.getName()).willReturn("name");
+        given(ageField.getName()).willReturn("age");
+
+        psiFieldsForConstructor.clear();
+        psiFieldsForSetters.clear();
+        allSelectedPsiFields.clear();
```
Comment (mjedynak) [inline src/test/java/pl/mjedynak/idea/plugins/builder/psi/BuilderPsiClassBuilderTest.java]: we're not reusing the fields, so we don't need to clear them

File: src/main/java/pl/mjedynak/idea/plugins/builder/psi/BuilderPsiClassBuilder.java [hunk 6]
```diff
@@ -130,8 +149,18 @@ private void appendConstructor(StringBuilder buildMethodText) {
...
+        appendSetMethods(buildMethodText, psiFieldsForSetters);
+        if (isInnerBuilder(builderClass)) {
+            Set<PsiField> fieldsSetViaAssignment = new HashSet<PsiField>(allSelectedPsiFields);
+            fieldsSetViaAssignment.removeAll(psiFieldsForSetters);
+            fieldsSetViaAssignment.removeAll(psiFieldsForConstructor);
+            appendAssignments(buildMethodText, fieldsSetViaAssignment);
+        }
+    }
+
+    private void appendSetMethods(StringBuilder buildMethodText, Collection<PsiField> fieldsBeSetViaSetter) {
```
Comment (mjedynak) [inline src/main/java/pl/mjedynak/idea/plugins/builder/psi/BuilderPsiClassBuilder.java]: 'fieldsToBeSet...'

**Comment by mjedynak** [comment]
👍, small issues fixed by myself

**Comment by blindpirate** [comment]
Cool. :-)
